### PR TITLE
Add field and check donation for phone verified 

### PR DESCRIPTION
**Pull request by Krishnag09**
The PR adds a new profile field to track donation, to act as a paywall for phone verification

closes the issue - https://github.com/Couchers-org/couchers/issues/4349

--->

**Backend checklist**
- [x] Formatted my code by running `ruff check --select I --fix . && ruff check . && ruff format .` in `app/backend`
- [x] Added tests for any new code or added a regression test if fixing a bug
- [x] All tests pass
- [x] Run the backend locally and it works
- [x] Added migrations if there are any database changes, rebased onto `develop` if necessary for linear migration history

**Web frontend checklist**
- [ ] Formatted my code with `make format`
- [ ] There are no warnings from `make lint`
- [ ] There are no console warnings when running the app
- [ ] Added any new components to storybook
- [ ] Added tests where relevant
- [ ] All tests pass
- [ ] Clicked around my changes running locally and it works
- [ ] Checked Desktop, Mobile and Tablet screen sizes

### CODE REVIEW

File: app/backend/src/couchers/models.py [hunk 1]
```diff
@@ -608,6 +610,21 @@ class StrongVerificationCallbackEvent(Base):
     iris_status = Column(String, nullable=False)
 
 
+class PhoneVerificationCallbackEvent(Base):
```
Comment (aapeliv) [inline app/backend/src/couchers/models.py]: I think this is unused?

File: app/backend/src/couchers/servicers/donations.py [hunk 2]
```diff
@@ -112,6 +112,10 @@ def Webhook(self, request, context):
                 # amount comes in cents
                 amount = int(data_object["amount"]) // 100
                 receipt_url = data_object["receipt_url"]
+
+                # may be check for amount to enable phone verify
+                session.add(User(has_donated=True))
```
Comment (aapeliv) [inline app/backend/src/couchers/servicers/donations.py]: This syntax creates a new user with just this flag set. The SQLalchemy ORM loads the user and will automatically update it so you just need to do `user.has_donated = True`

File: app/proto/api.proto [hunk 3]
```diff
@@ -265,6 +265,7 @@ message User {
   google.protobuf.BoolValue parking = 119;
   ParkingDetails parking_details = 120; // CommonMark without images
   google.protobuf.BoolValue camping_ok = 121;
+  google.protobuf.BoolValue has_donated = 123;
```
Comment (aapeliv) [inline app/proto/api.proto]: This probably shouldn't be returned to other users.

**Review by aapeliv — COMMENTED** [review]
I think there are two things that still need to be implemented:

1. It would be good to return the `has_donated` flag in GetAccountInfo (account.proto/account.py)
2. We need to somehow populate the `has_donated` flag when we set it up so that those who have already donated get it ticked off. I can help make this happen in the migration.

File: app/backend/src/tests/test_fixtures.py [hunk 4]
```diff
@@ -303,6 +303,11 @@ def invocation_metadata(self):
 
         session.commit()
 
+        if has_donated:
+            user.has_donated = True
+
+        assert user.has_donated == has_donated
+
```
Comment (aapeliv) [inline app/backend/src/tests/test_fixtures.py]: This is superfluous

File: app/media/src/media/server.py [hunk 5]
```diff
@@ -9,13 +9,12 @@
 import pyvips
 import sentry_sdk
 from flask import Flask, abort, request, send_file
+from media.crypto import verify_hash_signature
```
Comment (aapeliv) [inline app/media/src/media/server.py]: Please undo formatting in `app/media`

**Review by aapeliv — COMMENTED** [review]
1. Database migrations (update the `has_donated` flag based on existing data in the database) -- I can help with this: it's going to be complicated

This needs a database query to check who has already donated based on existing tables, so we set `has_donated=True` for those users.

2. Add a env var flag or constant in `constants.py` as to whether to enforce this or not. So we can update the frontend first.

File: app/backend/src/tests/test_fixtures.py (hunk 4 above)
Comment (aapeliv) [reply app/backend/src/tests/test_fixtures.py]: The `kwargs` should take care of this, no need to handle `has_donated` separately

**Comment by aapeliv** [comment]
@Krishnag09 : could you please rebase onto `develop` or merge `develop` into this branch, please?

Are you OK if I push a commit to fix up the migrations onto this branch? Other than that and minor formatting stuff (I can fix that up too), I think this is looking good to go!

**Comment by Krishnag09** [comment]
> 1. Database migrations (update the `has_donated` flag based on existing data in the database) -- I can help with this: it's going to be complicated
> 
> This needs a database query to check who has already donated based on existing tables, so we set `has_donated=True` for those users.
> 
> 2. Add a env var flag or constant in `constants.py` as to whether to enforce this or not. So we can update the frontend first.

do you still want me to do these?

**Comment by aapeliv** [comment]
> do you still want me to do these?

It's all done now in the commit I pushed. I think it just needs an update to the migration (to say down revision is `3ac39dcf3b5a`).

But let's wait a little bit before merging so I can do the frontend part first!

**Pull request merged**